pyuvdata/uvcal/ms_cal.py

```diff
@@ -0,0 +1,582 @@
+                for idx, cat_id in enumerate(self.phase_center_catalog):
+                    field_ids[self.phase_center_id_array == cat_id] = idx
+                field_ids = np.tile(np.repeat(field_ids, Nants_casa), self.Nspws)
+
+            # spw ids are on the outer most axis, so they just need to be repeated.
+            spw_id_array = np.repeat(np.arange(self.Nspws), Nants_casa * self.Ntimes)
+
+            # Set this always to zero for now -- not sure when its needed yet.
+            obs_id_array = np.zeros_like(spw_id_array)
+
+            # Calculate the number of times we need to duplicate records b/c of the way
+            # that CASA stores information.
+            ms.putcol("TIME", time_array)
+            ms.putcol("FIELD_ID", field_ids)
+            ms.putcol("SPECTRAL_WINDOW_ID", spw_id_array)
+            ms.putcol("ANTENNA1", ant_array)
+            ms.putcol("ANTENNA2", refant_array)
+            ms.putcol("INTERVAL", interval_arr)
+            ms.putcol("SCAN_NUMBER", scan_number_array)
+            ms.putcol("OBSERVATION_ID", obs_id_array)
+
+            # Alright, all the easy stuff is over, time to move on to the the heavy
+            # lifting, which we'll do spectral window by spectral window.
+            for count, spw_id in enumerate(self.spw_array):
+                # TODO: Change placeholder value
+                pol_order = ...
+                spw_sel, spw_nchan = spw_sel_dict[spw_id]
+                subarr_shape = (self.Ntimes, Nants_casa, spw_nchan, self.Njones)
+
+                qual_arr = np.ones(
+                    (self.Nants_data, spw_nchan, self.Ntimes, self.Njones), dtype=float
+                )
+                if self.quality_array is not None:
+                    qual_arr *= self.quality_array[:, spw_sel, :, :]
+                if self.total_quality_array is not None:
+                    qual_arr *= self.total_quality_array[spw_sel, :, :]
+
+                # We're going to leave a placeholder for SNR for now, since it's
+                # somewhat redundant and not totally clear how it's calculated.
+                data_dict = {
+                    cal_column: cal_array[:, spw_sel, :, :],
+                    "FLAG": self.flag_array[:, spw_sel, :, :],
+                    "PARAMERR": qual_arr,
+                    "SNR": np.zeros_like(qual_arr),
+                }
+                # N.b. (Karto): WEIGHT was totally unfilled in every example file I
+                # poked at that was generated via CASA. It's possible it's not actually
+                # used for anything or is otherwise vestigial, so leave it be for now.
+
+                for item in data_dict:
+                    subarr = data_dict[item]
+                    # data_form = ("Nants_data", "Nfreqs", "Ntimes", "Njones")
+                    # Order it so that it goes Ntimes -> Nants -> Nfreqs -> Njones
+                    subarr = np.transpose(subarr, [2, 0, 1, 3])
```

Comment:
The comments don't line up with axis number order in the transpose call. If both are correct, we'll need more commenting so a future reader isn't confused like I am.

Reply:
I think this is just an unclear comment -- I was using it to help me make sense of/tracking the mapping, but I can see how it's not super clear.

Reply:
Fixed in e5922d902a3fef3192b79753192ebcae05366859.